### PR TITLE
EQL: Add concat function

### LINKED ISSUE #55185: EQL: implement concat function
Part of #51556

Call `toString()` on all the non-null input arguments.

For the null behavior, we could go one of three ways:
* if any of the inputs are null, return null
* if all of the inputs are null, return null
* return an empty string, if all inputs are null

Update from 2020/04/23:
* We decided to go with the first option. All inputs are required, so if any are missing, we will return null. This will be more consistent with other functions, and make the strings within concat more consistently parseable.

### PR DESCRIPTION
**Pull request by rw-access**
Closes #55185

Adds variable argument `concat` function. This function will never return null, and only skips over null input parameters.

In #55185, I raise three different options for null handling. Currently, I'm choosing the first.
1. never return null
2. return null if any inputs are null
3. return null if all inputs are null

But I'm wondering if the second option (return null if any input is null) is the most appropriate and consistent with null handling. Having a coalesce function to make the user manually handle nulls seems like a better choice than just ignoring null inputs.

### CODE REVIEW

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Concat.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    @Override
+    protected TypeResolution resolveType() {
+        if (!childrenResolved()) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = TypeResolution.TYPE_RESOLVED;
+        int index = 0;
+        for (Expression value : values) {
+            resolution = isExact(value, sourceText(), ParamOrdinal.fromIndex(index));
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Concat.java:50]: Why only `isExact`? Shouldn't this be a string only type of value?

Comment (astefan) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Concat.java:50]: Also, this loop goes through all of `concat`'s values and returns the resolution of the last element in list. Shouldn't, maybe, stop at the first resolution that is `.unresolved()` and return that?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public final Pipe replaceChildren(List<Pipe> newChildren) {
+        return new ConcatFunctionPipe(source(), expression(), newChildren);
+    }
+
+    @Override
+    public final Pipe resolveAttributes(AttributeResolver resolver) {
+        List<Pipe> newValues = new ArrayList<>(values.size());
+        for (Pipe v: values) {
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java:36]:
```suggestion
        for (Pipe v : values) {
```

And the same change throughout the rest of this function's classes. Thanks.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        if (newValues == values) {
+            return this;
+        }
+
+        return replaceChildren(newValues);
+    }
+
+    @Override
+    public boolean supportedByAggsOnlyQuery() {
+        for (Pipe p: values) {
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java:49]:
```suggestion
        for (Pipe p : values) {
```

File: x-pack/plugin/eql/qa/common/src/main/resources/test_queries_supported.toml [hunk 4]
```diff
@@ -14,6 +14,16 @@ query = '''
 file where between(file_path, "dev", ".json", true) == "\\TestLogs\\something"
 '''
 
+[[queries]]
```
Comment (astefan) [inline x-pack/plugin/eql/qa/common/src/main/resources/test_queries_supported.toml:17]: Please, add more integration tests here:
- with one pattern only
- with an empty string pattern among other patterns
- with `null` patterns

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            if (p.supportedByAggsOnlyQuery() == false) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean resolved() {
+        for (Pipe p: values) {
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java:59]:
```suggestion
        for (Pipe p : values) {
```

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            if (p.resolved() == false) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public final void collectFields(QlSourceBuilder sourceBuilder) {
+        for (Pipe v: values) {
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/ConcatFunctionPipe.java:69]:
```suggestion
        for (Pipe v : values) {
```

**Review by astefan — COMMENTED** [review]
Looks good in general, but I would like to see more tests, please.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Concat.java (hunk 1 above)
Comment (rw-access) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Concat.java:50]: fixed in 9295e7920777ebca4c3d26cc51cedecdbf4869db

Comment (rw-access) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Concat.java:50]: also, this accepts non string parameters as well, and calls `.toString()` on them
the example uses `pid` which is a long
https://eql.readthedocs.io/en/latest/query-guide/functions.html#concat

File: x-pack/plugin/eql/qa/common/src/main/resources/test_queries_supported.toml (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/eql/qa/common/src/main/resources/test_queries_supported.toml:17]: With all nulls as well

**Review by matriv — COMMENTED** [review]
Seems good overall, I agree with @astefan about more tests with nulls and empty strings.

- Should `ConcatFunctionPipe` also have tests?
- Also missing the context maybe, shouldn't it operate on `text` too (inexact) ?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/planner/QueryFolderFailTests.java [hunk 7]
```diff
@@ -93,6 +93,14 @@ public void testCIDRMatchNonString() {
                 "line 1:15: argument of [cidrMatch(source_address, 12345)] must be [string], found value [12345] type [integer]", msg);
     }
 
+    public void testConcatWithInexact() {
+        VerificationException e = expectThrows(VerificationException.class,
+            () -> plan("process where concat(plain_text)"));
+        String msg = e.getMessage();
+        assertEquals("Found 1 problem\nline 1:15: [concat(plain_text)] cannot operate on first argument field of data type "
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/planner/QueryFolderFailTests.java:100]: Small improvement here: if it's possible for a one argument call to not use "first" in the error message, that would look more user friendly.

Also, I would like to see one or two tests where another argument different from the first doesn't fulfill the restriction.

**Review by astefan — APPROVED** [review]
LGTM. Left one more comment regarding the error message that gets logged when the fields are not of the required type.

**Review by matriv — APPROVED** [review]
LGTM

**Comment by rw-access** [comment]
Backport 389082033efe33b5a64c7607cd9900272dbdbfa1

**Pull request merged**